zeebe/broker/src/main/java/io/camunda/zeebe/broker/system/partitions/impl/MigrationSnapshotDirector.java

```diff
@@ -0,0 +1,164 @@
+      final AsyncSnapshotDirector asyncSnapshotDirector,
+      final ConcurrencyControl control,
+      final Duration scheduleDelay,
+      final HealthMonitor healthMonitor) {
+    snapshotDirector = asyncSnapshotDirector;
+    this.control = control;
+    this.scheduleDelay = scheduleDelay;
+    this.healthMonitor = healthMonitor;
+    healthReport = snapshotNotTaken();
+    healthMonitor.registerComponent(this);
+    LOG.debug("Initialized migration snapshot director. Scheduling snapshot");
+    scheduleSnapshot();
+  }
+
+  @Override
+  public void close() {
+    cancelScheduledSnapshot();
+    healthMonitor.removeComponent(this);
+  }
+
+  public void scheduleSnapshot() {
+    if (!snapshotTaken) {
+      LOG.debug("Scheduling snapshot for migration.");
+      forceSnapshotUntilSuccessful();
+    }
+  }
+
+  public void cancelScheduledSnapshot() {
+    if (runningSnapshot != null) {
+      runningSnapshot.cancel();
+      runningSnapshot = null;
+    }
+  }
+
+  @VisibleForTesting
+  public boolean isSnapshotTaken() {
+    return snapshotTaken;
+  }
+
+  @Override
+  public String componentName() {
+    return COMPONENT_NAME;
+  }
+
+  @Override
+  public HealthReport getHealthReport() {
+    return healthReport;
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    listeners.put(failureListener, Boolean.TRUE);
+  }
+
+  @Override
+  public void removeFailureListener(final FailureListener failureListener) {
+    listeners.remove(failureListener);
+  }
+
+  private void forceSnapshotUntilSuccessful() {
+    forceSnapshotUntilSuccessful(null);
+  }
+
+  private void forceSnapshotUntilSuccessful(final Throwable error) {
+    var shouldBeRetried = !snapshotTaken && runningSnapshot == null;
+    if (error != null) {
+      shouldBeRetried &=
+          error instanceof IOException
+              || (error instanceof CompletionException && error.getCause() instanceof IOException);
+    }
+    if (shouldBeRetried) {
+      runningSnapshot =
+          control.schedule(
+              scheduleDelay,
+              () -> {
+                forceSnapshot()
+                    .onComplete(
+                        (ignored, ex) -> {
+                          forceSnapshotUntilSuccessful(ex);
+                        },
+                        control);
+              });
+    }
+  }
+
+  private ActorFuture<Void> forceSnapshot() {
+    if (!snapshotTaken) {
+      return snapshotDirector
+          .forceSnapshot()
+          .andThen(
+              (snapshot, error) -> {
+                if (snapshot != null) {
+                  // reset the flag
+                  snapshotTaken = true;
+                  LOG.debug("Snapshot taken after migrations: {}", snapshot.getId());
+                  healthReport = HealthReport.healthy(this);
+                  notifyListeners();
+                } else {
+                  LOG.debug("Snapshot not taken after migrations, retrying in {}.", scheduleDelay);
```

Comment:
❓️As there is no back off (?), this could be called in a tight loop? So maybe trace it, or use a ThrottledLogger? Or is my assessment wrong?

Reply:
I will add exponential backoff with max ~5 seconds